src/cephadm/build.py

```diff
@@ -106,6 +117,24 @@ def _install_deps(tempdir):
     )
 
 
+def generate_version_file(versioning_vars, dest):
+    log.info("Generating version file")
+    log.debug("versioning_vars=%r", versioning_vars)
+    with open(dest, "w") as fh:
```

Comment:
for my clarity, this isn't currently actually compiling this into the binary, it's just writing out this `_version.py` file in the same dir we're outputting the binary and then `cephadm version` later will try to import it?

Reply:
No, we write out a file named `_version.py` into the directory (tempdir) we pass to the function `_compile` which puts it into the root of the zipapp. Here's an example using `zipinfo` to show what's inside the "compiled" binary:
```
$ zipinfo ./_out.0427/bin/cephadm 
Archive:  ./_out.0427/bin/cephadm
Zip file size: 86861 bytes, number of entries: 2
-rwxr-xr-x  2.0 unx   384610 b- defN 23-Apr-27 18:09 __main__.py
-rw-r--r--  2.0 unx      167 b- defN 23-Apr-27 18:09 _version.py
2 files, 384777 bytes uncompressed, 86621 bytes compressed:  77.5%
```

You can see `__main__.py` - that's cephadm.py - and the new `_version.py`.  The contents of the zip become part of the "python path" so we can import stuff from within the zip. Therefore we have a special top-level python module that contains our "build information" at `_version.py`.

Reply:
would it be better to create a `pyproject.toml` that could potentially be reused by other tools (e.g `setuptools` etc.)?

Reply:
I don't know if it would be worth it since those expect a more typical python packaging approach and ceph is being using cmake and does a lot of things uniquely. I like being able to give cephadm a version string more similar to the `ceph` command than your average python package.